kapitan/remoteinventory/fetch.py

```diff
@@ -0,0 +1,201 @@
+import os
+import tempfile
+import logging
+from collections import defaultdict
+from distutils.dir_util import copy_tree
+from functools import partial
+from shutil import copyfile, rmtree
+from git import Repo
+import hashlib
+
+from git import GitCommandError
+from kapitan.utils import make_request, unpack_downloaded_file
+from kapitan.errors import GitSubdirNotFoundError
+from kapitan.errors import GitFetchingError
+from kapitan import cached
+
+logger = logging.getLogger(__name__)
+
+
+def fetch_inventories(inventory_path, target_objs, pool):
+    """Parses through the 'inventory' parameter in target_objs to fetch the remote
+       inventories and stores it in a temporary directory before recursively
+       copying it to the output_path (relative to the inventory path)
+
+       :param inventory_path: default or user specified inventory path
+       :param target_objs: target objects
+       :param pool: pool object for multiprocessing
+
+       :return: None
+    """
+    temp_dir = tempfile.mkdtemp()
+
+    git_inventories = defaultdict(list)
+    http_inventories = defaultdict(list)
+
+    # To ensure no dublicate output path
+    inv_output_path = defaultdict(set)
+
+    for target_obj in target_objs:
+        try:
+            inventories = target_obj["inventories"]
+            for inv in inventories:
+                inv_type = inv["type"]
+                source_uri = inv["source"]
+                source_hash = hashlib.sha256(source_uri.encode())
+                # hashing the source and subdir together for git sources
+                # as different inventory items can have the same git uri
+                if "subdir" in inv:
+                    subdir = inv["subdir"]
+                    source_hash.update(subdir.encode())
+                if source_hash in cached.inv_sources:
+                    continue
+                # output_path is relative to inventory_path
+                # ./inventory by default
+                output_path = os.path.join(inventory_path, inv["output_path"])
+                if output_path in inv_output_path[source_uri]:
+                    continue
+                else:
+                    inv_output_path[source_uri].add(output_path)
+
+                if inv_type == "git":
+                    git_inventories[source_uri].append(inv)
+                elif inv_type in ("http", "https"):
+                    http_inventories[source_uri].append(inv)
+        except KeyError:
+            continue
```

Comment:
Add logger.debug() output in exceptions before continuing. It helps debugging. :)